[SE-0458] Adopt strict memory safety in the standard libraries

DougGregor opened the pull request:
With the acceptance of [SE-0458 "Opt-In Strict Memory Safety Checking"](https://swiftlang.github.io/swift-evolution/#?proposal=SE-0458) and its [enablement in the compiler](https://github.com/swiftlang/swift/pull/79645), adopt strict memory safety through most of the Standard Library. This involves passing the `-strict-memory-safety` flag to the compiler when building these standard library modules, and adding all of the necessary `@safe`/`@unsafe`/`unsafe` annotations to acknowledge uses of existing unsafe code.

There are a *lot* of `@unsafe` and `unsafe` annotations, hence the large diff. This is, fundamentally, because it's the Swift standard libraries that layer safe abstractions (arrays, dictionaries, strings, etc.) on top of unsafe primitives (raw UTF-8 bytes and such). The annotation here was largely mechanical, driven by compiler Fix-Its with some review and massaging.

DougGregor committed:
[SE-0458] Enable strict memory safety in the Swift standard library

DougGregor committed:
Adopt unsafe annotations throughout the standard library

compnerd commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -745,6 +745,16 @@ SupplementaryOutputPathsComputer::readSupplementaryOutputFileMap() const {
+        // if it was present on the command line.
+        auto &paths = outputPaths.back();
+        if (auto fixitPath = Args.getLastArg(options::OPT_emit_fixits_path)) {
+          if (paths.FixItsOutputPath.empty()) {
```
Why this check? The `if (auto ... = Args.getLastArg(options::OPT_emit_fixits_path))` should get just the last entry.

DougGregor replied:
Presumably, the supplementary outputs file could have an entry here, so we're saying that would take precedence over the command-line argument.

compnerd commented on stdlib/public/CMakeLists.txt:
```diff
@@ -18,7 +18,7 @@ endif()
   list(APPEND SWIFT_RUNTIME_SWIFT_COMPILE_FLAGS "-Xfrontend" "-sil-based-debuginfo")
 endif()
-
+  
```
Bleeding whitespace

compnerd replied:
Bleeding whitespace

tshortli commented on stdlib/public/core/ArrayBuffer.swift:
```diff
@@ -575,7 +575,7 @@ extension _ArrayBuffer {
   @inlinable @_alwaysEmitIntoClient
   internal func getAssociatedBuffer() -> _ContiguousArrayBuffer<Element>? {
-    let getter = unsafeBitCast(
+    let getter = unsafe unsafeBitCast(
```
We're going to need condfail mitigations in all the fragile functions, right?

DougGregor replied:
Yeah, I had an idea on that... since we have a swift-syntax tree for the inlined text, I think we can write a visitor that replaces `UnsafeExprSyntax` nodes with their argument.

DougGregor replied:
I'm glad my SwiftIfConfig rewrite can be useful :)

edymtt commented on stdlib/cmake/modules/SwiftSource.cmake:
```diff
@@ -1054,17 +1054,32 @@ function(_compile_swift_files
     set(copy_legacy_layouts_dep)
   endif()
 
+  # Produce a file that applies all Fix-Its, as a convenience.
```
It is not entirely clear to me from the commit message and the comments if the logic to emit the fixit will be dropped from this PR before merging, or if it will stay for future adoptions.

In the latter case,
* would it be possible to port this logic to the new build systems as well, so to maintain parity?
* could we gate this under a CMake `option`, so it is enabled at desk only when needed and we avoid landing those in packaged toolchains and SDKs?

DougGregor replied:
I think it's best if I drop this before merging. It's a tool to get us to enable strict memory safety throughout the standard libraries, but it's probably a bad idea to keep around.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Enable strict memory safety in the Distributed module

DougGregor committed:
Enable strict memory safety in the Synchronization module

DougGregor committed:
Update module-trace tests for strictly memory-safe standard libraries

DougGregor committed:
Make test robust against strict memory safety changes to the standard library

DougGregor committed:
Parsing and type checking for the definition of isolated conformances

Allow a conformance to be "isolated", meaning that it stays in the same
isolation domain as the conforming type. Only allow this for
global-actor-isolated types.

When a conformance is isolated, a nonisolated requirement can be
witnessed by a declaration with the same global actor isolation as the
enclosing type.

DougGregor committed:
Check consistency of isolation of conformances within other conformances

When a protocol conformance somehow depends on an isolated conformance, it
must itself be isolated to the same global actor as the conformance on
which it depends.

DougGregor committed:
[SE-0458] Enable unsafe expressions / attributes / for..in effects by default

With the acceptance of SE-0458, allow the use of unsafe expressions, the
@safe and @unsafe attributes, and the `unsafe` effect on the for..in loop
in all Swift code.

Introduce the `-strict-memory-safety` flag detailed in the proposal to
enable strict memory safety checking. This enables a new class of
feature, an optional feature (that is *not* upcoming or experimental),
and which can be detected via `hasFeature(StrictMemorySafety)`.

DougGregor committed:
[SE-0458] Disambiguate "unsafe" expression and for..in effect

Port over the disambiguation code we already had in the new Swift parser
to the existing C++ parser for the "unsafe" expression and for..in effect.

DougGregor committed:
Add a ChangeLog entry for SE-0458

DougGregor committed:
Enable strict memory safety in the SwiftOnoneSupport module

DougGregor committed:
Enable strict memory safety in the Concurrency module

DougGregor committed:
Enable strict memory safety in the Cxx/CxxStdlib modules

DougGregor committed:
[Serialization] Add the STRICT_MEMORY_SAFETY record type.

DougGregor commented:
Note that we have some serious stacking of PRs happening right now. The first five commits are covered by https://github.com/swiftlang/swift/pull/79645

DougGregor commented:
https://github.com/swiftlang/swift-syntax/pull/2978

@swift-ci please smoke test macOS

DougGregor commented:
https://github.com/swiftlang/swift-syntax/pull/2978

@swift-ci please smoke test Windows

eeckstein commented:
Just do double check that this (large) change has no impact on code generation:

@swift-ci apple silicon benchmark

DougGregor commented:
@eeckstein There should be zero impact on performance because there's no SIL-level differences for anything related to the strict memory safety checking. I think the results of that performance run look fine, but there are a few's ?'s I don't know how to interpret.

eeckstein commented:
Yes, it's completely fine. The "?" results are noise. If there would be a real impact, the results would look much different.
